cocos/2d/components/ui-opacity.ts

```diff
@@ -106,50 +127,93 @@ export class UIOpacity extends Component {
-                render.renderEntity.localOpacity = interruptOpacity;
+                render.renderEntity.localOpacity = parentOpacity;
             }
-            render.node._uiProps.localOpacity = render.renderEntity.localOpacity;
-            interruptOpacity = 1;
+            //No need for recursion here. Because it doesn't affect the capacity of the child nodes.
+            return;
         } else if (uiOp) {
             // there is a just UIOpacity but no UIRenderer on the node.
             // we should transport the interrupt opacity downward
-            interruptOpacity = interruptOpacity * uiOp.opacity / 255;
-            uiOp._setByParent = setByParent;
+            uiOp._parentOpacity = parentOpacity;
+            parentOpacity = uiOp._parentOpacity * uiOp.opacity / 255;
         }
+
         for (let i = 0; i < node.children.length; i++) {
-            UIOpacity.setEntityLocalOpacityDirtyRecursively(node.children[i], dirty || (interruptOpacity < 1), interruptOpacity, true);
+            UIOpacity.setEntityLocalOpacityDirtyRecursively(
+                node.children[i],
+                dirty || (parentOpacity < 1),
+                parentOpacity,
+                recursiveWithoutUiOp,
+            );
         }
     }
 
     @serializable
     protected _opacity = 255;
 
-    public onEnable (): void {
-        // If the ancestor node has a uiopacity component, it will be initialized when initializing 
-        // the uiopacity component of the ancestor node, and there is no need to initialize it again.
-        if (this._setByParent) {
-            return;
+    protected _getParentOpacity (node: Node): number {
+        if (node == null || !node.isValid) {
+            return 1;
         }
-        this.node._uiProps.localOpacity = this._opacity / 255;
-        this.setEntityLocalOpacityDirtyRecursively(true);
+        const render = node._uiProps.uiComp as UIRenderer;
+        const uiOp = node.getComponent<UIOpacity>(UIOpacity);
+        if (render && render.color) {
+            return 1;
+        } else if (uiOp) {
+            return uiOp._parentOpacity * (uiOp._opacity / 255);
+        }
+        return this._getParentOpacity(node.getParent()!);
     }
 
-    public onDisable (): void {
-        // If the ancestor node has a uiopacity component, it will be uninitialized when uninitializing 
-        // the uiopacity component of the ancestor node, and there is no need to uninitialize it again.
-        if (this._setByParent) {
+    protected _parentChanged (): void {
+        const parent = this.node.getParent();
+        let opacity = 1;
+        if (parent) {
+            this._parentOpacity = this._getParentOpacity(parent);
+            opacity = this._parentOpacity;
+        }
+        UIOpacity.setEntityLocalOpacityDirtyRecursively(this.node, true, opacity, false);
+    }
+
+    protected _setEntityLocalOpacityRecursively (opacity: number): void {
+        if (!JSB) {
+            return;
+        }
```

Comment:
Why return if not JSB?

Reply:
This is only necessary for native

Reply:
Could you please add some comment why only need in native?

Reply:
Because JSB's localOpacity value is present in the renderEntity, but non-JSB's are not.